src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs

```diff
@@ -0,0 +1,192 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using EnvDTE;
+using EnvDTE80;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.Shell;
+using Nuget.PackageIndex.Client.Analyzers;
+
+namespace Nuget.PackageIndex.VisualStudio
+{
+    /// <summary>
+    /// This filter is needed only until we enable Missing Package suggestions for all C# projects.
+    /// For now we limit this feature only to ProjectK scenarios.
+    /// </summary>
+    internal class DnxProjectFilter : IProjectFilter, IDisposable
+    {
+        private ConcurrentDictionary<string, string> FilesCache = new ConcurrentDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+        private bool _disposed;
+        private SolutionEvents _solutionEvents;
+        private object _cacheLock = new object();
+        private bool _solutionIsClosing;
+
+        public DnxProjectFilter()
+        {
+            var container = ServiceProvider.GlobalProvider.GetService<IComponentModel, SComponentModel>();
```

Comment:
It must be on UI thread (by Roslyn), but you are right, we have a helper to throw and I will add it there